syntax = "proto3";
package steam;
import "google/protobuf/timestamp.proto";

// Send and receive messages from a single Steam user
service MessageService {
    rpc SendChatMessage (MessageRequest) returns (SendMessageResult);
    rpc PollChatMessages (PollRequest) returns (stream ResponseMessage);
    rpc GetFriendsList (FriendsListRequest) returns (FriendsListResponse);
}

message MessageRequest {
    string targetId = 1;
    string message = 2;
    string sessionKey = 3;
}

message SendMessageResult {
    enum SendMessageResultCode {
        SUCCESS = 0;
        INVALID_SESSION_KEY = 1;
        INVALID_TARGET_ID = 2;
        INVALID_MESSAGE = 3;
        UNKNOWN_ERROR = 4;
    }

    bool success = 1;
    SendMessageResultCode reason = 2;
    string reasonStr= 3;
}

message PollRequest {
    string targetId = 1;
    string sessionKey = 2;
    optional google.protobuf.Timestamp lastTimestamp = 3;
}

message ResponseMessage {
    string senderId = 1;
    string message = 2;
    google.protobuf.Timestamp timestamp = 3;
}

message FriendsListRequest {
    string sessionKey = 1;
}

enum PersonaState {  // mirrors SteamUser.EPersonaState
    OFFLINE = 0;
    ONLINE = 1;
    BUSY = 2;
    AWAY = 3;
    SNOOZE = 4;
    LOOKING_TO_TRADE = 5;
    LOOKING_TO_PLAY = 6;
    INVISIBLE = 7;
}

message Persona {
    string id = 1;
    string name = 2;
    PersonaState personaState = 3;
}

message FriendsListResponse {
    repeated Persona friends = 1;
}